src/eos-updater-apply.c

```diff
@@ -73,41 +77,592 @@ get_test_osname (void)
+      return flatpak_installation_new_for_path (override, TRUE, cancellable, error);
+    }
+
+  return flatpak_installation_new_system (cancellable, error);
+}
+
+static GHashTable *
+determine_set_of_installed_applications (FlatpakInstallation *installation,
+                                         GCancellable        *cancellable,
+                                         GError              **error)
+{
+  g_autoptr(GPtrArray) refs = flatpak_installation_list_installed_refs (installation,
+                                                                        cancellable,
+                                                                        error);
+  GHashTable *set = NULL;
+  gsize i = 0;
+
+  if (!refs)
+    return NULL;
+
+  set = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+
+  for (; i < refs->len; ++i)
+    g_hash_table_add (set,
+                      g_strdup (flatpak_ref_get_name (g_ptr_array_index (refs, i))));
+
+  return set;
+}
+
+static GHashTable *
+determine_set_of_explicitly_uninstalled_applications_from_flatpak_log (GPtrArray *events,
+                                                                       GError    **error)
+{
+  GHashTable *table = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  gsize i = 0;
+
+  for (; i < events->len; ++i)
+    {
+      JsonObject *object = g_ptr_array_index (events, i);
+      const gchar *event = json_object_get_string_member (object, "event");
+      g_autoptr(FlatpakRef) ref = NULL;
+
+      if (g_strcmp0 (event, "uninstall") != 0)
+        continue;
+
+      ref = flatpak_ref_parse (json_object_get_string_member (object, "ref"),
+                               error);
+
+      if (!ref)
+        return NULL;
+
+      g_hash_table_add (table, g_strdup (flatpak_ref_get_name (ref)));
+    }
+
+  return table;
+}
+
+static GPtrArray *
+transactions_from_log_file (GFile  *transaction_log_file,
+                            GError **error)
+{
+  g_autoptr(JsonParser) parser = NULL;
+  g_autoptr(GPtrArray) transactions = NULL;
+  g_autofree gchar *contents = NULL;
+  gchar *contents_sep = NULL;
+  gchar *line = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  if (!g_file_get_contents (g_file_get_path (transaction_log_file),
+                            &contents,
+                            NULL,
+                            &local_error))
+    {
+      if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
+        }
+
+      /* If the file wasn't found, assume that no applications have been
+       * explicitly uninstalled */
+      return g_ptr_array_new ();
+    }
+
+  transactions = g_ptr_array_new_full (0, (GDestroyNotify) json_object_unref);
+  parser = json_parser_new ();
+
+  contents_sep = contents;
+  while ((line = strsep (&contents_sep, "\n")) != NULL)
```

Comment:
Any way to refactor this as a `for` to make the termination condition more explicit?

Reply:
Would be nice to use the same loop style we use elsewhere (right now that would mean splitting out a `GStrv`, I have an optional suggestion to use `GDataInputStream` instead)